HF - consolidate shards of safetensors files to full tensors in finish step

ankitageorge opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #156705
* #154743

Title - we can consolidate the shards to a full tensors, optionally behind a flag, in the finish step of DCP.save
also adds the thread count argument which is configurable for users, before we were just using the default of 1.
Re-creating https://github.com/pytorch/pytorch/pull/155940 bc it got into a bad detached state

Differential Revision: [D77231774](https://our.internmc.facebook.com/intern/diff/D77231774/)

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta

ankitageorge committed:
HF - consolidate shards of safetensors files to full tensors in finish step

Title - we can consolidate the shards to a full tensors, optionally behind a flag, in the finish step of DCP.save
also adds the thread count argument which is configurable for users, before we were just using the default of 1.
Re-creating https://github.com/pytorch/pytorch/pull/155940 bc it got into a bad detached state

Differential Revision: [D77231774](https://our.internmc.facebook.com/intern/diff/D77231774/)

[ghstack-poisoned]

ankitageorge committed:
Update on "HF - consolidate shards of safetensors files to full tensors in finish step"


Title - we can consolidate the shards to a full tensors, optionally behind a flag, in the finish step of DCP.save
also adds the thread count argument which is configurable for users, before we were just using the default of 1.
Re-creating https://github.com/pytorch/pytorch/pull/155940 bc it got into a bad detached state

Differential Revision: [D77231774](https://our.internmc.facebook.com/intern/diff/D77231774/)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

ankitageorge committed:
Update on "HF - consolidate shards of safetensors files to full tensors in finish step"


Title - we can consolidate the shards to a full tensors, optionally behind a flag, in the finish step of DCP.save
also adds the thread count argument which is configurable for users, before we were just using the default of 1.
Re-creating https://github.com/pytorch/pytorch/pull/155940 bc it got into a bad detached state

Differential Revision: [D77231774](https://our.internmc.facebook.com/intern/diff/D77231774/)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

saumishr commented on torch/distributed/checkpoint/hf_storage.py:
```diff
@@ -136,8 +153,16 @@ def write_data(
+        if self.save_sharded and not self.consolidated_output_path:
+            logger.info("Not consolidating sharded checkpoint in finish step.")
             return
+        if self.save_sharded:
```
As we discussed, current set up is the following:
1. Save_sharded: False -> Rank-0 checkpointing
2. Save_sharded: True -> consolidated path provided -> consolidated checkpoint
3. save_sharded: True -> consolidated path not provided -> distributed checkpoint

This is slightly confusing since the consolidated checkpoint is also not sharded. 
We should probably update the save_sharded config name to save_distributed to avoid the sharding confusion.

ankitageorge committed:
Update on "HF - consolidate shards of safetensors files to full tensors in finish step"


Title - we can consolidate the shards to a full tensors, optionally behind a flag, in the finish step of DCP.save
also adds the thread count argument which is configurable for users, before we were just using the default of 1.
Re-creating https://github.com/pytorch/pytorch/pull/155940 bc it got into a bad detached state

Differential Revision: [D77231774](https://our.internmc.facebook.com/intern/diff/D77231774/)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

ankitageorge committed:
Update on "HF - consolidate shards of safetensors files to full tensors in finish step"


Title - we can consolidate the shards to a full tensors, optionally behind a flag, in the finish step of DCP.save
also adds the thread count argument which is configurable for users, before we were just using the default of 1.
Re-creating https://github.com/pytorch/pytorch/pull/155940 bc it got into a bad detached state

Differential Revision: [D77231774](https://our.internmc.facebook.com/intern/diff/D77231774/)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

saumishr commented on torch/distributed/checkpoint/hf_storage.py:
```diff
@@ -136,8 +154,16 @@ def write_data(
     def finish(self, metadata: Metadata, results: list[list[WriteResult]]) -> None:
-        if self._save_sharded:
+        if self.save_distributed and not self.consolidated_output_path:
+            logger.info("Not consolidating sharded checkpoint in finish step.")
```
It would be worth adding a documentation here explaining different scenarios and why the metadata write is also being skipped.

ankitageorge committed:
Update on "HF - consolidate shards of safetensors files to full tensors in finish step"


Title - we can consolidate the shards to a full tensors, optionally behind a flag, in the finish step of DCP.save
also adds the thread count argument which is configurable for users, before we were just using the default of 1.
Re-creating https://github.com/pytorch/pytorch/pull/155940 bc it got into a bad detached state

Differential Revision: [D77231774](https://our.internmc.facebook.com/intern/diff/D77231774/)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

ankitageorge commented:
@pytorchmergebot merge

jithunnair-amd commented:
@pytorchbot revert -c nosignal -m "Introduced errors on ROCm CI. These errors were not seen on CUDA CI because CUDA CI docker images do not have safetensors installed and the test silently passes. Test tries to read from a /tmp location, but that file was not written"

HUD page showing first error occurrence: https://hud.pytorch.org/hud/pytorch/pytorch/c811f41cf58238b49e9c2f0a9fa00dcc471b2e10/1?per_page=50&name_filter=periodic-rocm-mi300&mergeEphemeralLF=true

ROCm failure after this PR's commit in trunk: https://github.com/pytorch/pytorch/actions/runs/15987254081/job/45095060576

Error: `FileNotFoundError: No such file or directory: "/tmp/tmp1om06w0l/consolidated/model-00001-of-00001.safetensors"`

https://github.com/pytorch/pytorch/blob/1ab1ab38a04e8ee852ff27eb8ae4989662511965/test/distributed/checkpoint/test_hf_safetensor_e2e.py#L127
ROCM: https://github.com/pytorch/pytorch/blob/e3351b3ddff06c90b2786b23312f80fda2ddb4a6/.ci/docker/build.sh#L248
CUDA: https://github.com/pytorch/pytorch/blob/e3351b3ddff06c90b2786b23312f80fda2ddb4a6/.ci/docker/build.sh#L105

jithunnair-amd commented:
@ankitageorge It seems there are quite a few PRs that will need to be reverted to revert this PR. We have filed a PR to install safetensors for *all* CI docker images: https://github.com/pytorch/pytorch/pull/158674.... Please issue a forward fix for this issue, thanks!
